lib/esConnector.js

```diff
@@ -1266,6 +1276,309 @@ ESConnector.prototype.updateOrCreate = function updateOrCreate(modelName, data,
+          return Promise.resolve(res);
+        }
+      },
+      function(err){
+        console.trace(err.message);
+        if (typeof err !== 'undefined' && err !== null) {
+          if (typeof done === 'function') {
+            done(err, null);
+          } else {
+            return Promise.reject(err);
+          }
+        }
+      }
+    );
+};
+
+/**
+ * ES bulk insert + index
+ * @param modelName {String} - required, model that is associated with this connector
+ * @param dataList {Array} - required, list of documents/models, that gets inserted and indexed
+ * @param params {Object} - passed, through to the bulk API
+ * @param done {Function} - required, allows async execution
+ * @returns {Promise} OR {undefined} if callback (done) is provided
+ * @value: {[String]} - return value is an array of object IDs
+ * @throws {Error} - if argument rules are violated
+ *
+ * NOTE: if params has a field called `body`, it will be ignored. Instead dataList is used
+ */
+ESConnector.prototype.bulkCreate = function bulkCreate(modelName, dataList, params, done) {
+  var self = this;
+  var functionName = 'bulkCreate';
+  if (self.debug) {
+    log('ESConnector.prototype.' + functionName, modelName, dataList, params);
+  }
+
+  if (Array.isArray(dataList) === false) {
+    var typeError = new Error('invalid type for dataList');
+    if (typeof done === 'function') {
+      return done(typeError);
+    } else {
+      return Promise.reject(typeError);
+    }
+  }
+
+  if (typeof params === 'function') {
+    done = params;
+    params = {};
+  } else if (typeof params !== 'object' || params === null) {
+    params = {};
+  }
+
+  var idName = self.idName(modelName);
+  var actionName = 'index';
+
+  // NOTE: incl. the names of type and index, if configured
+  var defaultParams = self.addDefaults(modelName, functionName);
+  var bulkParams = _.assign({}, defaultParams, params);
+
+  bulkParams.body = dataList.reduce(function(dataBulk, data){
+    var idValue = self.getIdValue(modelName, data);
+
+    var clonedData = _.assign({}, data);
+
+    delete clonedData[idName];
+    var action = {};
+    action[actionName] = { _id: idValue };
+    dataBulk.push(action, clonedData);
+
+    return dataBulk;
+  }, [ /* dataBulk */ ]);
+
+  return self
+    .bulk(modelName, bulkParams)
+    .then(
+      function(res){
+        log('ESConnector.prototype.bulkCreate', 'successful executed in: ' + res.took + 'ms');
+        
+        var result = res.items.map(function(item){
+          return item[actionName]._id;
+        });
+
+        if (typeof done === 'function') {
+          done(null, result);
+        } else {
+          return Promise.resolve(result);
+        }
+      },
+      function(err){
+        console.trace(err.message);
+        if (typeof err !== 'undefined' && err !== null) {
+          if (typeof done === 'function') {
+            done(err, null);
+          } else {
+            return Promise.reject(err);
+          }
+        }
```

Comment:
*tl;dr: as of the ES JS API specification, this is never going to happen, hence it is not necessary to handle.*

Generally, in Node.js-land the error-first-callback pattern applies. In reality, It often depends on the framework's implementation (node standard lib vs third party packages). The first parameter can either be `undefined` or `null`, but if and only if no error occurred in the callee context.

In this specific context, we rely on the behaviour of the JS client implementation of Elasticsearch's [Bulk API](https://www.elastic.co/guide/en/elasticsearch/client/javascript-api/current/api-reference.html). The required package version already supports *Promises* by skipping the callback function parameter. This is just a syntactical difference. How errors occur is the same. It boils down to either:
(A) in a callback pattern there is just one function, that is executed in both cases: error occurred or not, hence the check `if (typeof err !== 'undefined' && err !== null)` is required to distinguish between those two cases
(B) promises provide two different function, each for every case (error occurred or not). Here the error check might be redundant, because if the *rejected* function is called it already implies that `typeof err !== 'undefined'` nor `err !== null`.

All API endpoints added in this PR support both, callback-pattern and the Promise syntax. Therefore, a *thenable* object needs to be returned if `typeof done !== 'function'`, which means the API is used as a Promise, and in case `done` is a function it needs to be called according to the callback pattern.

I hope this clears things up.